File: Command/c#/Command.cs
interface ICommand
{
    void Positive();
    void Negative();
}

class Conveyor
{
    public void On() => Console.WriteLine("the pipeline is running");

    public void Off() => Console.WriteLine("the conveyor is stopped");

    public void SpeedIncrease() => Console.WriteLine("increased conveyor speed");

    public void SpeedDecrease() => Console.WriteLine("reduced conveyor speed");
}

class ConveyorWorkCommand : ICommand
{
    private Conveyor _conveyor;

    public ConveyorWorkCommand(Conveyor conveyor) => _conveyor = conveyor;

    public void Positive() => _conveyor.On();

    public void Negative() => _conveyor.Off();
}

class ConveyorAdjustCommand : ICommand
{
    private Conveyor _conveyor;

    public ConveyorAdjustCommand(Conveyor conveyor) => _conveyor = conveyor;

    public void Positive() => _conveyor.SpeedIncrease();
    
    public void Negative() => _conveyor.SpeedDecrease();
}

class Multipult
{
    private List<ICommand?> _commands;
    private Stack<ICommand> _history;

    public Multipult()
    {
        _commands = new List<ICommand>() {null, null};
        _history = new Stack<ICommand>();
    }

    public void SetCommand(int button, ICommand command) => _commands[button] = command;

    public void PressOn(int button)
    {
        _commands[button].Positive();
        _history.Push(_commands[button]);
    }

    public void PressCancel()
    {
        if (_history.Count != 0)
            _history.Pop().Negative();
    }
}

class Program
{
    static void Main()
    {
        Conveyor conveyor = new Conveyor();

        Multipult multipult = new Multipult();
        multipult.SetCommand(0, new ConveyorWorkCommand(conveyor));
        multipult.SetCommand(1, new ConveyorAdjustCommand(conveyor));
        
        multipult.PressOn(0);
        multipult.PressOn(1);
        multipult.PressCancel();
        multipult.PressCancel();
    }
}

File: Command/python/Command.py
import abc
from typing import List, Deque


class ICommand(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def positive(self):
        pass
    
    
    @abc.abstractmethod
    def negative(self):
        pass


class Conveyor:

    def on(self):
        print('the pipeline is running')


    def off(self):
        print('the conveyor is stopped')


    def speed_increase(self):
        print('increased conveyor speed')


    def speed_decrease(self):
        print('reduced conveyor speed')


class ConveyorWorkCommand(ICommand):

    def __init__(self, conveyor: Conveyor):
        self.conveyor: Conveyor = conveyor


    def positive(self):
        self.conveyor.on()


    def negative(self):
        self.conveyor.off()


class ConveyorAdjustCommand(ICommand):

    def __init__(self, conveyor: Conveyor):
        self.conveyor: Conveyor = conveyor


    def positive(self):
        self.conveyor.speed_increase()


    def negative(self):
        self.conveyor.speed_decrease()


class Multipult:

    def __init__(self):
        self.__commands: List[ICommand] = [None, None]
        self.__history: Deque[ICommand] = []


    def set_command(self, button: int, command: ICommand):
        self.__commands[button] = command


    def press_on(self, button: int):
        self.__commands[button].positive()
        self.__history.append(self.__commands[button])
        

    def press_cancel(self):
        if len(self.__history) != 0:
            self.__history.pop().negative()


if __name__ == '__main__':
    conveyor = Conveyor()

    multipult = Multipult()
    multipult.set_command(0, ConveyorWorkCommand(conveyor))
    multipult.set_command(1, ConveyorAdjustCommand(conveyor))

    multipult.press_on(0)
    multipult.press_on(1)
    multipult.press_cancel()
    multipult.press_cancel()